gossipd/gossip.c

```diff
@@ -1667,31 +1667,47 @@ static struct io_plan *conn_init(struct io_conn *conn, struct reaching *reach)
 	return io_connect(conn, &ai, connection_out, reach);
 }
 
-static void try_connect(struct reaching *reach)
+static void try_reach_peer(struct daemon *daemon, const struct pubkey *id,
+			   bool master_needs_response)
 {
 	struct addrhint *a;
 	int fd;
+	struct reaching *reach;
+	u8 *msg;
+	struct peer *peer = find_peer(daemon, id);
 
-	/* Already succeeded somehow? */
-	if (find_peer(reach->daemon, &reach->id)) {
-		status_trace("Already reached %s, not retrying",
-			     type_to_string(tmpctx, struct pubkey, &reach->id));
-		tal_free(reach);
+	if (peer) {
+		status_debug("try_reach_peer: have peer %s",
+			     type_to_string(tmpctx, struct pubkey, id));
+		if (master_needs_response) {
```

Comment:
Just tell the master, no need to disambiguate here.

Reply:
Might be cleaner to split this into two functions.  One caller is because of an explicit connect, which needs a response, one for an implicit (important peer retry), which doesn't.

But there are several things which can go wrong, so the return from such a function would be complex:
1. Peer already connected.
2. Currently trying to reach peer.
3. Some kind of socket/connect error.

Hence the "tell master" flag, though it's ugly.